cpp/src/arrow/dataset/file_test.cc

```diff
@@ -353,6 +361,153 @@ TEST_F(TestFileSystemDataset, WriteProjected) {
+    return MakeVectorIterator(std::move(fragments));
+  };
+};
+
+TEST_F(TestFileSystemDataset, WritePersistOrder) {
+  // Test for GH-26818
+  auto format = std::make_shared<IpcFileFormat>();
+  FileSystemDatasetWriteOptions write_options;
+  write_options.file_write_options = format->DefaultWriteOptions();
+  write_options.base_dir = "root";
+  write_options.partitioning = std::make_shared<HivePartitioning>(schema({}));
+  write_options.basename_template = "{i}.feather";
+
+  // The Mock dataset delays emitting the first fragment, which test sequenced output of
+  // scan node
+  auto dataset = std::make_shared<MockDataset>(schema({field("f0", int32())}));
+
+  // The delay scalar function delays some batches of all fragments, which tests implicit
+  // ordering
+  auto delay_func = std::make_shared<compute::ScalarFunction>("delay", compute::Arity(1),
+                                                              compute::FunctionDoc());
+  compute::ScalarKernel delay_kernel;
+  delay_kernel.exec = delay;
+  delay_kernel.signature = compute::KernelSignature::Make({int32()}, boolean());
+  ASSERT_OK(delay_func->AddKernel(delay_kernel));
+  ASSERT_OK(compute::GetFunctionRegistry()->AddFunction(delay_func));
+
+  for (bool preserve_order : {true, false}) {
+    ASSERT_OK_AND_ASSIGN(auto scanner_builder, dataset->NewScan());
+    ASSERT_OK(scanner_builder->UseThreads(true));
+    ASSERT_OK(
+        scanner_builder->Filter(compute::call("delay", {compute::field_ref("f0")})));
+    ASSERT_OK_AND_ASSIGN(auto scanner, scanner_builder->Finish());
+
+    auto fs = std::make_shared<fs::internal::MockFileSystem>(fs::kNoTime);
+    write_options.filesystem = fs;
+    write_options.preserve_order = preserve_order;
+
+    ASSERT_OK(FileSystemDataset::Write(write_options, scanner));
+
+    // Read the file back out and verify the order
+    ASSERT_OK_AND_ASSIGN(auto dataset_factory, FileSystemDatasetFactory::Make(
+                                                   fs, {"root/0.feather"}, format, {}));
+    ASSERT_OK_AND_ASSIGN(auto written_dataset, dataset_factory->Finish(FinishOptions{}));
+    ASSERT_OK_AND_ASSIGN(scanner_builder, written_dataset->NewScan());
+    ASSERT_OK(scanner_builder->UseThreads(false));
+    ASSERT_OK_AND_ASSIGN(scanner, scanner_builder->Finish());
+    ASSERT_OK_AND_ASSIGN(auto actual, scanner->ToTable());
+    TableBatchReader reader(*actual);
+    std::shared_ptr<RecordBatch> batch;
+    ASSERT_OK(reader.ReadNext(&batch));
+    int32_t prev = -1;
+    auto out_of_order = false;
+    while (batch != nullptr) {
+      const auto* values = batch->column(0)->data()->GetValues<int32_t>(1);
+      for (int row = 0; row < batch->num_rows(); ++row) {
+        int32_t value = values[row];
+        if (value <= prev) {
+          out_of_order = true;
+        }
+        prev = value;
+      }
+      ASSERT_OK(reader.ReadNext(&batch));
+    }
+    ASSERT_EQ(!out_of_order, preserve_order);
```

Comment:
As mentioned in my other comment about the unordered-ness, this could result in false-negative I guess?

I.e., we should assert that no out-of-order batches are observed if `preserve_order == true`, though it might not always be a true-positive (the batches are out-of-order underlying and it was our order-preserving logic that made them in order). But when `preserve_order == false`, I don't think we can simply assert that out-of-order always happens, given the probabilistic of the unordered-ness.

Reply:
Out-of-order probability with `reserve_order == false` is very high, virtually 100%.

See https://github.com/apache/arrow/pull/44470/files#r2079049038.

Reply:
Thanks for the explanation.

However my point is more of that, when `preserve_order == false`, we don't really have to assert ooo happens, because we don't really guarantee ordering nor ooo, do we? Though in practice, it's likely always ooo and the risk of false-negative is near zero, but we don't even have to take the risk at the first place, right?

Reply:
So I'm suggesting simply changing the assertion to:

```suggestion
    ASSERT_TRUE(!preserve_order || !out_of_order); // No out-of-order is allowed if preserve_order is specified true.
```

Reply:
If I understand @EnricoMi correctly - it doesn't hurt to actually check ooo when `preserve_order == false` as it guarantees our generated test data is ooo before sorting it.

Reply:
My bias is that not having false-negative should be prioritized over testing the test setup. Given the near 100% ooo probability in practice, I'm not standing in the way if other reviewers think it's OK. Thanks.